iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+        // only accept the request if it has a free slot in its active view, otherwise it will refuse the request."
+        match details.priority {
+            Priority::High => {
+                self.add_active(from, details.data, Priority::High, now, io);
+            }
+            Priority::Low if !self.active_is_full() => {
+                self.add_active(from, details.data, Priority::Low, now, io);
+            }
+            _ => {}
+        }
+    }
+
+    fn peer_info(&self, id: &PA) -> Option<PeerInfo<PA>> {
+        self.peer_data.get(id).map(|data| PeerInfo {
+            id: *id,
+            data: data.clone(),
+        })
+    }
+
+    fn insert_peer_info(&mut self, peer_info: PeerInfo<PA>, io: &mut impl IO<PA>) {
+        let old = self.peer_data.remove(&peer_info.id);
+        let same = matches!(old, Some(old) if old == peer_info.data);
+        if !same {
+            io.push(OutEvent::PeerData(peer_info.id, peer_info.data.clone()));
+        }
+        self.peer_data.insert(peer_info.id, peer_info.data);
+    }
+
+    fn on_shuffle(&mut self, from: PA, shuffle: Shuffle<PA>, io: &mut impl IO<PA>) {
+        if shuffle.ttl.expired() {
+            let len = shuffle.nodes.len();
+            for node in shuffle.nodes {
+                self.add_passive(node.id, node.data, io);
+            }
+            let nodes = self
+                .passive_view
+                .shuffled_max(len, &mut self.rng)
+                .into_iter()
+                .map(|id| self.peer_info(&id).unwrap());
+            let message = Message::ShuffleReply(ShuffleReply {
+                nodes: nodes.collect(),
+            });
+            io.push(OutEvent::SendMessage(shuffle.origin, message));
+        } else if let Some(node) = self
+            .active_view
+            .pick_random_without(&[&shuffle.origin, &from], &mut self.rng)
+        {
+            let message = Message::Shuffle(Shuffle {
+                origin: shuffle.origin,
+                nodes: shuffle.nodes,
+                ttl: shuffle.ttl.next(),
+            });
+            io.push(OutEvent::SendMessage(*node, message));
+        }
+    }
+
+    fn on_shuffle_reply(&mut self, message: ShuffleReply<PA>, io: &mut impl IO<PA>) {
+        for node in message.nodes {
+            self.add_passive(node.id, node.data, io);
+        }
+    }
+
+    fn on_disconnect(&mut self, peer: PA, details: Disconnect, io: &mut impl IO<PA>) {
+        self.pending_neighbor_requests.remove(&peer);
+        self.remove_active(&peer, details.respond, io);
+        if details.alive {
+            if let Some(data) = self.peer_data.remove(&peer) {
+                self.add_passive(peer, data, io);
+            }
+        } else {
+            self.passive_view.remove(&peer);
+        }
+    }
+
+    fn handle_shuffle_timer(&mut self, io: &mut impl IO<PA>) {
+        if let Some(node) = self.active_view.pick_random(&mut self.rng) {
+            let active = self.active_view.shuffled_without_max(
```

Comment:
when I see `shuffled_without_max` I read _shuffle and remove the maximum_.  Maybe `shuffled_without_and_capped` or `shuffled_without_capped` would work? Open to other options?